app/src/main/java/org/schabi/newpipe/RouterActivity.java

```diff
@@ -257,70 +258,110 @@ private void showUnsupportedUrlDialog(final String url) {
-                    break;
-                case 0:
-                    handleChoice(showInfoKey);
-                    break;
-                default:
-                    showDialog(choices);
-                    break;
-            }
-        } else if (selectedChoiceKey.equals(showInfoKey)) {
-            handleChoice(showInfoKey);
-        } else if (selectedChoiceKey.equals(downloadKey)) {
-            handleChoice(downloadKey);
-        } else {
+
+        final ChoiceAvailabilityChecker choiceChecker = new ChoiceAvailabilityChecker(
+                getChoicesForService(currentService, currentLinkType),
+                preferences.getString(getString(R.string.preferred_open_action_key),
+                        getString(R.string.preferred_open_action_default)));
+
+        // Check for non-player related choices
+        if (choiceChecker.isAvailableAndSelected(
+                R.string.show_info_key,
+                R.string.download_key,
+                R.string.add_to_playlist_key)) {
+            handleChoice(choiceChecker.getSelectedChoiceKey());
+            return;
+        }
+        // Check if the choice is player related
+        if (choiceChecker.isAvailableAndSelected(
+                R.string.video_player_key,
+                R.string.background_player_key,
+                R.string.popup_player_key)) {
+
+            final String selectedChoice = choiceChecker.getSelectedChoiceKey();
+
             final boolean isExtVideoEnabled = preferences.getBoolean(
                     getString(R.string.use_external_video_player_key), false);
             final boolean isExtAudioEnabled = preferences.getBoolean(
                     getString(R.string.use_external_audio_player_key), false);
-            final boolean isVideoPlayerSelected = selectedChoiceKey.equals(videoPlayerKey)
-                    || selectedChoiceKey.equals(popupPlayerKey);
-            final boolean isAudioPlayerSelected = selectedChoiceKey.equals(backgroundPlayerKey);
-
-            if (currentLinkType != LinkType.STREAM) {
-                if (isExtAudioEnabled && isAudioPlayerSelected
-                        || isExtVideoEnabled && isVideoPlayerSelected) {
-                    Toast.makeText(this, R.string.external_player_unsupported_link_type,
-                            Toast.LENGTH_LONG).show();
-                    handleChoice(showInfoKey);
-                    return;
-                }
+            final boolean isVideoPlayerSelected =
+                    selectedChoice.equals(getString(R.string.video_player_key))
+                            || selectedChoice.equals(getString(R.string.popup_player_key));
+            final boolean isAudioPlayerSelected =
+                    selectedChoice.equals(getString(R.string.background_player_key));
+
+            if (currentLinkType != LinkType.STREAM
+                    && (isExtAudioEnabled && isAudioPlayerSelected
+                    || isExtVideoEnabled && isVideoPlayerSelected)
```

Comment:
Parenthesis

Reply:
``&&`` binds stronger than ``||`` and I know that (also most Java devs should probably also know that...).

However - as I have to admit - this is not very clearly readable. 

Fixed.